### PR TITLE
Fix typing in `pl.callbacks.timer`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7037 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @carmocca @awaelchli

### CODE REVIEW

File: pytorch_lightning/callbacks/timer.py [hunk 1]
```diff
@@ -164,8 +164,10 @@ def on_load_checkpoint(
         self._offset = time_elapsed.get(RunningStage.TRAINING.value, 0)
 
     def _check_time_remaining(self, trainer: "pl.Trainer") -> None:
+        assert self._duration is not None
         should_stop = self.time_elapsed() >= self._duration
         should_stop = trainer.training_type_plugin.broadcast(should_stop)
+        should_stop = cast(bool, should_stop)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/timer.py:170]:
```suggestion
        should_stop = cast(bool, trainer.training_type_plugin.broadcast(should_stop))
```

**Review by kaushikb11 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/timer.py [hunk 2]
```diff
@@ -164,8 +164,9 @@ def on_load_checkpoint(
         self._offset = time_elapsed.get(RunningStage.TRAINING.value, 0)
 
     def _check_time_remaining(self, trainer: "pl.Trainer") -> None:
+        assert self._duration is not None
         should_stop = self.time_elapsed() >= self._duration
-        should_stop = trainer.training_type_plugin.broadcast(should_stop)
+        should_stop = cast(bool, trainer.training_type_plugin.broadcast(should_stop))
```
Comment (carmocca) [inline pytorch_lightning/callbacks/timer.py:169]: This should not be necessary, the fix should be in broadcast

where the correct signature update is

```python
TBroadcast = TypeVar("TBroadcast")

def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
```

Comment (ananthsub) [reply pytorch_lightning/callbacks/timer.py:169]: @four4fish do you recall why the above was changed^?

Comment (awaelchli) [reply pytorch_lightning/callbacks/timer.py]: @carmocca updated

Comment (carmocca) [reply pytorch_lightning/callbacks/timer.py]: LGTM, resolving

**Pull request merged**